File: stepik-union/src/main/java/org/stepik/core/utils/ProductGroup.java
package org.stepik.core.utils;

import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;

/**
 * @author meanmail
 */
public enum ProductGroup {
    UNKNOWN_GROUP, IDEA(Product.IDEA_CE, Product.IDEA_ULTIMATE), PYCHARM(Product.PYCHARM_CE, Product.PYCHARM_PRO);

    private static Map<Product, ProductGroup> map;
    private final Product[] products;

    ProductGroup(Product... products) {
        this.products = products;
    }

    public static ProductGroup of(Product product) {
        initMap();
        return map.getOrDefault(product, UNKNOWN_GROUP);
    }

    private static void initMap() {
        if (map == null) {
            map = new HashMap<>();

            Arrays.stream(values())
                    .forEach(productGroup -> Arrays.stream(productGroup.products)
                            .forEach(product -> map.put(product, productGroup)));
        }
    }

    @Override
    public String toString() {
        return String.format("%s %s", name(), Arrays.toString(products));
    }
}

File: stepik-union/src/main/java/org/stepik/core/utils/ProductGroup.kt
package org.stepik.core.utils


enum class ProductGroup constructor(vararg val products: Product) {
    UNKNOWN_GROUP, IDEA(Product.IDEA_CE, Product.IDEA_ULTIMATE), PYCHARM(Product.PYCHARM_CE, Product.PYCHARM_PRO);

    override fun toString(): String {
        return "$name $products"
    }

    companion object {

        private val map: Map<Product, ProductGroup> by lazy {
            values().map { group ->
                group.products.map { it to group }
            }.flatMap { it }.toMap()
        }

        fun of(product: Product): ProductGroup {
            return map[product] ?: UNKNOWN_GROUP
        }
    }
}
